### PR TITLE
mptcp: add notsent_lowat support

### PR DESCRIPTION
**Pull request by pabeni**
based on epoll_out_edge_notsent_lowat.pkt, with some adaptation to cope with mptcp specifics.

Note that this will cause some random failures around line 43.

The root cause is that the packetdrill core tries to aggregate eligible TCP packets and can replace the wire packet with the aggregate one to fit the drill under test.

The problem is that the aggregation support is partial/buggy: packets with the same tcp options  len, but different tcp options contents will be considered eligible for aggregation, and the aggregate packets will carry the options for the 1st pkt on the wire.

The above works e.g. with tcp stamp options, but does not work with mptcp: packets with different DSS are aggregated carrying the DSS from the 1st one. Note that this is different from what the GRO engine is doing and attempts to mimic what the TCP stack coalescing is doing. With the major difference that such coalescing merges correctly different DSS.

TL;DR: to solve the failures for good non trivial changes to the core are required, but sharing the drill early to somewhat covers the pending kernel patches

### CODE REVIEW

File: gtests/net/mptcp/syscalls/sockopt_SENT_LOWAT.pkt [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// connect() function, connection initiated by the kernel
```
Comment (matttbe) [inline gtests/net/mptcp/syscalls/sockopt_SENT_LOWAT.pkt]: about the location, should we not move it to the `sockopts` dir, instead of the `syscalls` one?

Comment (matttbe) [inline gtests/net/mptcp/syscalls/sockopt_SENT_LOWAT.pkt]: I'm not sure I understand the comment here. Should we remove it?

File: gtests/net/mptcp/syscalls/sockopt_SENT_LOWAT.pkt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+// Now this ack opens up substantial space and we will send EPOLLOUT.
++0.0      <    .  1:1(0)            ack 16001 win 8000 <dss dack8=16001                             nocs>
++0.0      >   P.  16001:20001(4000) ack 1              <dss dack4=1 dsn8=16001  ssn=16001 dll=4000  nocs, nop, nop>
++0.0      >   P.  20001:24001(4000) ack 1              <dss dack4=1 dsn8=20001  ssn=20001 dll=4000  nocs, nop, nop>
++0.0   epoll_wait(5, {events=EPOLLOUT, fd=4}, 1, 0) = 1
+
+// EPOLLET prevents a second EPOLLOUT.
++0.0   epoll_wait(5, {events=0, ptr=0}, 1, 0) = 0
+
+`sysctl -w net.ipv4.tcp_wmem="4096 262144 4194304"`
```
Comment (matttbe) [inline gtests/net/mptcp/syscalls/sockopt_SENT_LOWAT.pkt]: (it should not be needed when the tests are launched with `run_all.py` in a dedicated netns, but it doesn't hurt)

File: gtests/net/mptcp/syscalls/sockopt_SENT_LOWAT.pkt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// The ack does not open up enough space to send EPOLLOUT to the user.
++0.0      >  P.  1:4001(4000)       ack 1              <dss dack4=1 dsn8=1      ssn=1     dll=4000  nocs, nop,nop>
++0.0      >  P.  4001:8001(4000)    ack 1              <dss dack4=1 dsn8=4001   ssn=4001  dll=4000  nocs, nop,nop>
+
+// Now epoll_wait should not return any event because there is no space.
++0.0   epoll_wait(5, {events=0, ptr=0}, 1, 0) = 0
+
++0.0      <   .  1:1(0)             ack 8001  win 8000 <dss dack8=8001                              nocs>
++0.0      >  P. 8001:12001(4000)    ack 1              <dss dack4=1 dsn8=8001   ssn=8001  dll=4000  nocs, nop, nop>
++0.0      >  P. 12001:16001(4000)   ack 1              <dss dack4=1 dsn8=12001  ssn=12001 dll=4000  nocs, nop, nop>
```
Comment (matttbe) [inline gtests/net/mptcp/syscalls/sockopt_SENT_LOWAT.pkt]: should we add a comment here about the random failures? e.g.

    Packetdrill might try to coalesce packets here, but breaking MPTCP options.
    See commit message + https://github.com/multipath-tcp/mptcp_net-next/issues/476

Just to have a reminder when looking at failures for this test (without looking at the commit message)

**Review by matttbe — CHANGES_REQUESTED** [review]
Hi Paolo,

Thank you for the new test!

It looks good to me, just a few questions, nothing important, I can do the modifications myself if you prefer.

**Comment by matttbe** [comment]
I just applied the suggested modifications, see [here](https://github.com/multipath-tcp/packetdrill/compare/4d67bf945b9f6b0e65dfdab4e6013a03557f35fa..1902aa7c869c2a63bba4e17918624c3232ea6e32) (+ the modification of the commit message to copy what is in the PR message).
